src/samplers/interrupt/mod.rs

```diff
@@ -0,0 +1,150 @@
+            fatal!("failed to initialize interrupt sampler");
+        } else {
+            error!("failed to initialize interrupt sampler");
+        }
+    }
+
+    fn common(&self) -> &Common {
+        &self.common
+    }
+
+    fn common_mut(&mut self) -> &mut Common {
+        &mut self.common
+    }
+
+    fn sampler_config(&self) -> &dyn SamplerConfig<Statistic = Self::Statistic> {
+        self.common.config().samplers().interrupt()
+    }
+
+    async fn sample(&mut self) -> Result<(), std::io::Error> {
+        if let Some(ref mut delay) = self.delay() {
+            delay.tick().await;
+        }
+
+        if !self.sampler_config().enabled() {
+            return Ok(());
+        }
+
+        debug!("sampling");
+        self.register();
+
+        self.sample_interrupt().await?;
+
+        Ok(())
+    }
+
+    fn summary(&self, _statistic: &Self::Statistic) -> Option<Summary> {
+        let max = crate::common::hardware_threads().unwrap_or(1024) * SECOND;
+        Some(Summary::histogram(
+            max,
+            3,
+            Some(self.general_config().window()),
+        ))
+    }
+}
+
+impl Interrupt {
+    async fn sample_interrupt(&self) -> Result<(), std::io::Error> {
+        let file = File::open("/proc/interrupts").await?;
+        let reader = BufReader::new(file);
+        let mut lines = reader.lines();
+
+        let mut result = HashMap::new();
+
+        let mut cores: Option<usize> = None;
+        while let Some(line) = lines.next_line().await? {
+            let parts: Vec<&str> = line.split_whitespace().collect();
+            if cores.is_none() {
+                cores = Some(parts.len());
+                continue;
+            }
+            let mut sum = 0;
+            for i in 1..cores.unwrap() {
+                sum += parts.get(i).unwrap_or(&"0").parse().unwrap_or(0);
+            }
+            let stat = match parts.get(0) {
+                Some(&"NMI:") => InterruptStatistic::NonMaskable,
+                Some(&"LOC:") => InterruptStatistic::LocalTimer,
+                Some(&"SPU:") => InterruptStatistic::Spurious,
+                Some(&"PMI:") => InterruptStatistic::PerformanceMonitoring,
+                Some(&"RES:") => InterruptStatistic::Rescheduling,
+                Some(&"TLB:") => InterruptStatistic::TlbShootdowns,
+                Some(&"TRM:") => InterruptStatistic::ThermalEvent,
+                Some(&"MCE:") => InterruptStatistic::MachineCheckException,
+                _ => match parts.last() {
+                    Some(&"timer") => InterruptStatistic::Timer,
+                    Some(&"rtc0") => InterruptStatistic::RealTimeClock,
+                    Some(label) => {
+                        if label.len() >= 3 {
+                            if label[0..3] == *"mlx" || label[0..3] == *"eth" {
+                                InterruptStatistic::Network
+                            } else if label.len() >= 4 && label[0..4] == *"nvme" {
```

Comment:
same here, in the case there's multiple NVMe devices, do you want to differentiate them ?

Reply:
Currently, we export network and disk metrics in aggregate as opposed to per-device. This enables easier queries against the collected telemetry at the cost of less detail. If we'd like to start exporting per-device telemetry, we'll want to address it in multiple samplers. For now, I think the system-level aggregate will be sufficient, but can be convinced otherwise.